src/range.cr

```diff
@@ -392,12 +392,127 @@ struct Range(B, E)
       raise ArgumentError.new("Can't sample an open range")
     end
 
-    return super unless n == 1
+    if n < 0
+      raise ArgumentError.new "Can't sample negative number of elements"
+    end
+
+    # For a range of integers we can do much better
+    {% if B < Int && E < Int %}
+      min = self.begin
+      max = self.end
+
+      if (exclusive? && max <= min) || (!exclusive? && max < min)
+        raise ArgumentError.new "Invalid range for rand: #{self}"
+      end
+
+      max -= 1 if self.exclusive?
+
+      available = max - min + 1
+
+      # When a big chunk of elements is going to be needed, it's
+      # faster to just traverse the entire range than hitting
+      # a lot of duplicates because or random.
+      if n >= available // 4
+        return super
+      end
+
+      possible = Math.min(n, available)
+
+      # If we must return all values in the range...
+      if possible == available
+        result = Array(B).new(possible)
+        each { |value| result << value }
+        result.shuffle!(random)
+        return result
+      end
+
+      range_sample(n, random)
+    {% elsif B < Float && E < Float %}
+      min = self.begin
+      max = self.end
+
+      if (exclusive? ? max <= min : max < min)
+        raise ArgumentError.new "Invalid range for rand: #{self}"
+      end
+
+      if min == max
+        return [min]
+      end
+
+      range_sample(n, random)
+    {% else %}
+      case n
+      when 0
+        [] of B
+      when 1
+        [sample(random)]
+      else
+        super
+      end
+    {% end %}
+  end
+
+  # :nodoc:
+  def sample_old(n : Int, random = Random::DEFAULT)
+    {% if B == Nil || E == Nil %}
+      {% raise "Can't sample an open range" %}
+    {% end %}
 
-    if empty?
+    if self.begin.nil? || self.end.nil?
+      raise ArgumentError.new("Can't sample an open range")
+    end
+
+    if n < 0
+      raise ArgumentError.new "Can't sample negative number of elements"
+    end
+
+    case n
+    when 0
       [] of B
-    else
+    when 1
       [sample(random)]
+    else
+      raise ArgumentError.new("Can't sample negative number of elements") if n < 0
+
+      # Unweighted reservoir sampling:
+      # https://en.wikipedia.org/wiki/Reservoir_sampling#Simple_algorithm
+      # "Algorithm L" does not provide any performance improvements on Enumerable,
+      # because it is not possible to discard multiple elements at once
+
+      ary = Array(B).new(n)
+      return ary if n == 0
+
+      each_with_index do |elem, i|
+        if i < n
+          ary << elem
+        else
+          j = random.rand(i + 1)
+          if j < n
+            ary.to_unsafe[j] = elem
+          end
+        end
+      end
+
+      ary.shuffle!(random)
+    end
+  end
+
+  private def range_sample(n, random)
+    if n <= 16
```

Comment:
Can we use a constant (like `SMALL_ARRAY_SIZE`) here instead of a magic number?

Reply:
@straight-shoota isn't there an existing conversation on this PR talking about this? And FWIW I think @asterite has defended the choice quite well there.

Reply:
It has been to long... 🙈